File: frontend/src/api/sc/purchase/receive/model/__tests__/printReceiveSheetBo.test.ts
import { describe, expect, it } from 'vitest';

import type { PrintReceiveSheetBo } from '../printReceiveSheetBo';

const printData = {
  id: 'receive-1',
  code: 'RK202607050001',
  scId: 'sc-1',
  scCode: 'SC001',
  scName: '示例仓库',
  supplierId: 'supplier-1',
  supplierCode: 'SUP001',
  supplierName: '示例供应商',
  purchaserId: 'user-1',
  purchaserName: '张三',
  orderDate: '2026-07-05',
  paymentDate: '2026-07-06',
  receiveDate: '2026-07-07',
  purchaseOrderId: 'purchase-order-1',
  purchaseOrderCode: 'CG202607050001',
  totalNum: 10,
  totalGiftNum: 1,
  totalAmount: 120,
  paidAmount: 100,
  unpaidAmount: 20,
  description: '示例备注',
  createBy: '张三',
  createTime: '2026-07-05 10:00:00',
  updateBy: '李四',
  updateTime: '2026-07-05 11:00:00',
  approveBy: '王五',
  approveTime: '2026-07-05 12:00:00',
  status: 2,
  refuseReason: '',
  settleStatus: 1,
  details: [
    {
      id: 'receive-detail-1',
      productId: 'product-1',
      productCode: 'P0001',
      productName: '示例商品',
      skuCode: 'SKU001',
      externalCode: 'SP001',
      orderNum: 10,
      receiveNum: 10,
      unitId: 'unit-1',
      unitName: '袋',
      conversionRate: 1,
      businessNum: 10,
      taxPrice: 12,
      purchasePrice: 12,
      taxAmount: 120,
      receiveAmount: 120,
      isGift: false,
      taxRate: 13,
      description: '明细备注',
      orderNo: 1,
      purchaseOrderDetailId: 'purchase-order-detail-1',
      productionDate: '2026-07-01',
    },
  ],
} satisfies PrintReceiveSheetBo;

describe('采购入库打印响应模型', () => {
  it('覆盖完整采购入库字段和明细字段', () => {
    expect(printData.purchaseOrderId).toBe('purchase-order-1');
    expect(printData.totalGiftNum).toBe(1);
    expect(printData.settleStatus).toBe(1);
    expect(printData.details[0].productionDate).toBe('2026-07-01');
    expect(printData.details[0].receiveNum).toBe(10);
  });
});

File: backend/xingyun-sc/src/test/java/com/lframework/xingyun/sc/bo/purchase/receive/PrintReceiveSheetBoTest.java
package com.lframework.xingyun.sc.bo.purchase.receive;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.lframework.starter.web.core.utils.ApplicationUtil;
import com.lframework.xingyun.sc.dto.purchase.PurchaseProductDto;
import com.lframework.xingyun.sc.dto.purchase.receive.ReceiveSheetFullDto;
import com.lframework.xingyun.sc.service.purchase.PurchaseOrderService;
import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;
import org.testng.Assert;
import org.testng.annotations.Test;
import org.springframework.context.ApplicationContext;

/**
 * 采购入库打印模型字段测试。
 */
class PrintReceiveSheetBoTest {

  /**
   * 确认采购入库打印主表包含完整业务字段和已有展示字段。
   */
  @Test
  void shouldExposeCompleteReceiveSheetFields() {
    Set<String> fieldNames = getFieldNames(PrintReceiveSheetBo.class);
    Set<String> expectedFields = new HashSet<>(Arrays.asList(
        "id", "code", "scId", "supplierId", "purchaserId", "orderDate",
        "purchaseOrderId", "paymentDate", "receiveDate", "totalNum",
        "totalGiftNum", "totalAmount", "paidAmount", "unpaidAmount", "description",
        "createBy", "createTime", "updateBy", "updateTime", "approveBy", "approveTime",
        "status", "refuseReason", "settleStatus", "details", "scCode", "scName",
        "supplierCode", "supplierName", "purchaserName", "purchaseOrderCode"));

    Assert.assertTrue(fieldNames.containsAll(expectedFields),
        "采购入库打印主表缺少字段：" + difference(expectedFields, fieldNames));
  }

  /**
   * 确认采购入库打印明细包含原始业务字段和已有展示字段。
   */
  @Test
  void shouldExposeCompleteReceiveSheetDetailFields() {
    Set<String> fieldNames = getFieldNames(PrintReceiveSheetBo.OrderDetailBo.class);
    Set<String> expectedFields = new HashSet<>(Arrays.asList(
        "id", "productId", "orderNum", "unitId", "unitName", "conversionRate",
        "businessNum", "taxPrice", "taxAmount", "isGift", "taxRate", "description",
        "orderNo", "purchaseOrderDetailId", "productionDate", "productCode",
        "productName", "skuCode", "externalCode", "receiveNum", "purchasePrice",
        "receiveAmount", "inquiryProduct"));

    Assert.assertTrue(fieldNames.containsAll(expectedFields),
        "采购入库打印明细缺少字段：" + difference(expectedFields, fieldNames));
  }

  /**
   * 确认询价商品标识能从收货明细 DTO 原样映射到打印明细。
   */
  @Test
  void shouldMapInquiryProductToPrintDetail() throws ReflectiveOperationException {
    synchronized (ApplicationUtil.class) {
      ApplicationContext originalApplicationContext = getApplicationContext();
      ApplicationContext applicationContext = mock(ApplicationContext.class);
      PurchaseOrderService purchaseOrderService = mock(PurchaseOrderService.class);
      when(applicationContext.getBean(PurchaseOrderService.class)).thenReturn(purchaseOrderService);
      PurchaseProductDto product = new PurchaseProductDto();
      product.setCode("P001");
      product.setName("询价商品");
      when(purchaseOrderService.getPurchaseById("product-1")).thenReturn(product);
      new ApplicationUtil().setApplicationContext(applicationContext);

      try {
        ReceiveSheetFullDto.OrderDetailDto inquiryDetail = new ReceiveSheetFullDto.OrderDetailDto();
        inquiryDetail.setProductId("product-1");
        inquiryDetail.setInquiryProduct(true);
        ReceiveSheetFullDto.OrderDetailDto normalDetail = new ReceiveSheetFullDto.OrderDetailDto();
        normalDetail.setProductId("product-1");
        normalDetail.setInquiryProduct(false);

        Assert.assertTrue(
            new PrintReceiveSheetBo.OrderDetailBo(inquiryDetail).getInquiryProduct());
        Assert.assertFalse(
            new PrintReceiveSheetBo.OrderDetailBo(normalDetail).getInquiryProduct());
      } finally {
        new ApplicationUtil().setApplicationContext(originalApplicationContext);
      }
    }
  }

  /**
   * 获取测试开始前的全局 Spring 上下文。
   *
   * @return 当前全局 Spring 上下文
   * @throws ReflectiveOperationException 反射读取失败
   */
  private ApplicationContext getApplicationContext() throws ReflectiveOperationException {
    Field field = ApplicationUtil.class.getDeclaredField("APPLICATION_CONTEXT");
    field.setAccessible(true);
    return (ApplicationContext) field.get(null);
  }

  /**
   * 确认测试字段集合覆盖业务 DTO 当前全部非静态字段，防止新增字段遗漏。
   */
  @Test
  void shouldKeepPrintModelAlignedWithReceiveSheetDto() {
    Set<String> printRootFields = getFieldNames(PrintReceiveSheetBo.class);
    Set<String> dtoRootFields = getFieldNames(ReceiveSheetFullDto.class);
    Set<String> printDetailFields = getFieldNames(PrintReceiveSheetBo.OrderDetailBo.class);
    Set<String> dtoDetailFields = getFieldNames(ReceiveSheetFullDto.OrderDetailDto.class);

    Assert.assertTrue(printRootFields.containsAll(dtoRootFields),
        "打印主表未覆盖业务 DTO 字段：" + difference(dtoRootFields, printRootFields));
    Assert.assertTrue(printDetailFields.containsAll(dtoDetailFields),
        "打印明细未覆盖业务 DTO 字段：" + difference(dtoDetailFields, printDetailFields));
  }

  /**
   * 获取指定类型声明的非静态字段名称。
   *
   * @param type 类型
   * @return 字段名称集合
   */
  private Set<String> getFieldNames(Class<?> type) {
    return Arrays.stream(type.getDeclaredFields())
        .filter(field -> !Modifier.isStatic(field.getModifiers()))
        .map(Field::getName)
        .collect(Collectors.toSet());
  }

  /**
   * 计算期望字段集合中缺失的字段。
   *
   * @param expectedFields 期望字段
   * @param actualFields 实际字段
   * @return 缺失字段
   */
  private Set<String> difference(Set<String> expectedFields, Set<String> actualFields) {
    Set<String> difference = new HashSet<>(expectedFields);
    difference.removeAll(actualFields);
    return difference;
  }
}
